src/main/scala/com/databricks/spark/csv/CsvRelation.scala

```diff
@@ -147,18 +147,25 @@ case class CsvRelation protected[spark] (
   override def buildScan(requiredColumns: Array[String]): RDD[Row] = {
     val schemaFields = schema.fields
     val requiredFields = StructType(requiredColumns.map(schema(_))).fields
-    val isTableScan = requiredColumns.isEmpty || schemaFields.deep == requiredFields.deep
-    if (isTableScan) {
+    val shouldTableScan = schemaFields.deep == requiredFields.deep
+    val safeRequiredFields = if (dropMalformed) {
+      // If `dropMalformed` is enabled, then it needs to parse all the values
+      // so that we can decide which row is malformed.
+      requiredFields ++ schemaFields.filterNot(requiredFields.contains(_))
```

Comment:
Because of the order of columns. The required columns can be in a different order with the original schema.
If you look through the code below, you will notice I slice the row by `take()`.